README.md
# byteranges

A JavaScript `multipart/byteranges` parser library.

## Overview

This library parses `multipart/byteranges` response body of [HTTP range requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests).

The Range HTTP request header indicates the part of a document that the server should return. Several parts can be requested with one Range header at once, and the server may send back these ranges in a multipart document.

## Install

```
npm install byteranges
```

## Usage

```js
const { parse, getContentRange, BodyPart } = require('../build/byteranges');
const http = require('http');

http.get('http://www.columbia.edu/~fdc/picture-of-something.jpg', {
    headers: {
        Range: 'bytes=1-10,101-150'
    }
}, res => {
    const { statusCode } = res;
    if (statusCode !== 206) {
        res.resume();
        return;
    }

    const chunks = [];
    res.on('data', chunk => {
        chunks.push(chunk);
    });
    res.on('end', () => {
        const body = Buffer.concat(chunks);
        const contentType = res.headers['content-type'];

        if (!/multipart\/byteranges/i.test(contentType)) {
            const range = getContentRange(res.headers['content-range']);
            const part = new BodyPart(range, body, contentType);
            console.log(JSON.stringify(part));
            return;
        }

        const match = /boundary=(\w+)/i.exec(contentType);
        if (match === null) {
            return;
        }

        const boundary = match[1];
        const parts = parse(body, boundary);
        console.log(JSON.stringify(parts));
    });
});
```

Using [superagent](https://github.com/ladjs/superagent):

```js
const { parse, getContentRange, BodyPart } = require('../build/byteranges');
const superagent = require('superagent');

superagent
    .get('http://www.columbia.edu/~fdc/picture-of-something.jpg')
    .set('Range', 'bytes=1-10,101-150')
    .buffer(true)
    .parse(superagent.parse['application/octet-stream'])
    .end((err, res) => {
        if (err) {
            console.log(err);
            return;
        }

        if (res.status !== 206) {
            return;
        }

        const contentType = res.headers['content-type'];
        if (!/multipart\/byteranges/i.test(contentType)) {
            const range = getContentRange(res.headers['content-range']);
            const part = new BodyPart(range, res.body, contentType);
            console.log(JSON.stringify(part));
            return;
        }

        const match = /boundary=(\w+)/i.exec(contentType);
        if (match !== null) {
            const boundary = match[1];
            const parts = parse(res.body, boundary);
            console.log(JSON.stringify(parts));
        }
    });
```

CHANGELOG.md
# byteranges

## 1.1.0 (2023-02-12)

- Fix return type of parse function (parts cannot be unsatisfied ranges).
- Add isSatisfiableRange, isCompleteLengthRange, and isCompleteLengthRangeParts type guard functions.

## 1.0.1 (2023-01-21)

- Fix package.json.

## 1.0.0 (2023-01-21)

- Initial release.


test/byteranges.spec.ts
import fs from 'fs/promises';
import chai, { expect } from 'chai';
import chaiBytes from 'chai-bytes';
import { InclRange, getContentRange, parse, isCompleteLengthRangeParts, isCompleteLengthRange, ContentRange, isSatisfiableRange } from '../src/byteranges';

chai.use(chaiBytes);

describe('Byteranges', function () {
    it('should parse content-range when the complete length is known', function () {
        const header = 'bytes 42-1233/1234';
        const range = getContentRange(header);
        expect(range).to.be.an('object');
        expect(range.unit).to.equals('bytes');
        expect(range.range).to.be.an('object');
        expect(range.range).to.be.an.instanceOf(InclRange);
        expect((range.range as InclRange).start).to.eq(42);
        expect((range.range as InclRange).end).to.eq(1233);
        expect(range.length).to.equals(1234);
    });

    it('should parse content-range when the complete length is unknown', function () {
        const header = 'bytes 42-1233/*';
        const range = getContentRange(header);
        expect(range).to.be.an('object');
        expect(range.unit).to.equals('bytes');
        expect(range.range).to.be.an('object');
        expect(range.range).to.be.an.instanceOf(InclRange);
        expect((range.range as InclRange).start).to.eq(42);
        expect((range.range as InclRange).end).to.eq(1233);
        expect(range.length).to.equals('*');
    });

    it('should parse content-range with unsatisfied range', function () {
        const header = 'bytes */1234';
        const range = getContentRange(header);
        expect(range).to.be.an('object');
        expect(range.unit).to.equals('bytes');
        expect(range.range).to.equals('*');
        expect(range.length).to.equals(1234);
    });

    it('should throw an exception', function () {
        expect(getContentRange).to.throw();
    });

    it('should parse multipart/byteranges body', async function () {
        const body = await fs.readFile('test/byteranges');
        const parts = parse(body, '00000000000000000002');
        expect(parts).to.be.an('array').of.length(2);
        expect(parts[0].octets).to.equalBytes([255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255]);
        expect(parts[1].octets).to.equalBytes([
            105, 110, 116, 101, 114, 108, 97, 99, 101, 100, 61, 48, 32, 98, 108, 117,
            114, 97, 121, 95, 99, 111, 109, 112, 97, 116, 61, 48, 32, 99, 111, 110,
            115, 116, 114, 97, 105, 110, 101, 100, 95, 105, 110, 116, 114, 97, 61, 48,
            32, 98, 102, 114, 97, 109, 101, 115, 61, 51, 32, 98, 95, 112, 121, 114,
            97, 109, 105, 100, 61, 50, 32, 98, 95, 97, 100, 97, 112, 116, 61, 49,
            32, 98, 95, 98, 105, 97, 115, 61, 48, 32, 100, 105, 114, 101, 99, 116,
            61, 49, 32, 119, 101
        ]);
        expect(isCompleteLengthRangeParts(parts));
    });

    it('should assert that range is satisfiable', function () {
        const range = new ContentRange('bytes', new InclRange(0, 1), 1);
        expect(isSatisfiableRange(range)).to.be.true;
    });

    it('should assert that range is not satisfiable', function () {
        const range = new ContentRange('bytes', '*', '*');
        expect(isSatisfiableRange(range)).to.be.false;
    });

    it('should assert that range is complete', function () {
        const range = new ContentRange('bytes', new InclRange(0, 1), 1);
        expect(isCompleteLengthRange(range)).to.be.true;
    });
    
    it('should assert that range is not complete', function () {
        const range = new ContentRange('bytes', new InclRange(0, 1), '*');
        expect(isCompleteLengthRange(range)).to.be.false;
    });
});


examples/http.js
const { parse, getContentRange, BodyPart } = require('../build/byteranges');
const http = require('http');

http.get('http://www.columbia.edu/~fdc/picture-of-something.jpg', {
    headers: {
        Range: 'bytes=1-10,101-150'
    }
}, res => {
    const { statusCode } = res;
    if (statusCode !== 206) {
        res.resume();
        return;
    }

    const chunks = [];
    res.on('data', chunk => {
        chunks.push(chunk);
    });
    res.on('end', () => {
        const body = Buffer.concat(chunks);
        const contentType = res.headers['content-type'];

        if (!/multipart\/byteranges/i.test(contentType)) {
            const range = getContentRange(res.headers['content-range']);
            const part = new BodyPart(range, body, contentType);
            console.log(JSON.stringify(part));
            return;
        }

        const match = /boundary=(\w+)/i.exec(contentType);
        if (match === null) {
            return;
        }

        const boundary = match[1];
        const parts = parse(body, boundary);
        console.log(JSON.stringify(parts));
    });
});


src/byteranges.ts
import { Buffer } from 'node:buffer';


/**
 * `InclRange` objects are used to represent inclusive ranges.
 */
export class InclRange {
    constructor(public readonly start: number, public readonly end: number) { }
}

export type InclRangeOrStar = InclRange | '*';

export type CompleteLengthOrStar = number | '*';


/**
 * `ContentRange` objects are used to represent content ranges.
 */
export class ContentRange<T extends InclRangeOrStar = InclRangeOrStar, U extends CompleteLengthOrStar = CompleteLengthOrStar> {
    constructor(public readonly unit: string, public readonly range: T, public readonly length: U) { }
}


/**
 * `BodyPart` objects are used to represent body parts.
 */
export class BodyPart<T extends InclRangeOrStar = InclRangeOrStar, U extends CompleteLengthOrStar = CompleteLengthOrStar> {
    constructor(public readonly range: ContentRange<T, U>, public readonly octets: Buffer, public readonly type?: string) { }
}

const crlf = Buffer.from([0x0D, 0x0A]);
const crlfcrlf = Buffer.concat([crlf, crlf]);
const crvre = /(\w+)\s+(?:(\d+)-(\d+)\/(\d+|\*)|(\*)\/(\d+))/i;
const crkvre = /content-range:\s+(\w+)\s+(?:(\d+)-(\d+)\/(\d+|\*)|(\*)\/(\d+))/i;
const ctkvre = /content-type:\s+(.+)/i;

function getContentRangeInternal(match: RegExpExecArray): ContentRange {
    let unit: string = match[1];
    let range: InclRangeOrStar;
    let length: CompleteLengthOrStar;

    if (match[5] === '*') {
        range = '*';
        length = +match[6];
    } else if (match[4] === '*') {
        range = new InclRange(+match[2], +match[3]);
        length = '*';
    } else {
        range = new InclRange(+match[2], +match[3]);
        length = +match[4];
    }

    return new ContentRange(unit, range, length);
}

/**
 * Parses a HTTP Content-Range header `value`.
 * Returns a `ContentRange` instance.
 * 
 * Throws if the format of the supplied `value` is invalid.
 * 
 * @param value HTTP Content-Range header value.
 * @returns A `ContentRange` instance.
 */
export function getContentRange(value: string): ContentRange {
    const match = crvre.exec(value);

    if (match === null) {
        throw new Error('Invalid Content-Range value');
    }

    return getContentRangeInternal(match);
}

function getPartContentRange(headers: string): ContentRange {
    const match = crkvre.exec(headers);

    if (match === null) {
        throw new Error('Content-Range header not found');
    }

    return getContentRangeInternal(match);
}

function getPartContentType(headers: string): string | undefined {
    const match = ctkvre.exec(headers);

    return match === null ? undefined : match[1];
}


/**
 * Type guard function that narrows the supplied `ContentRange` instance to a `ContentRange<InclRange>` instance.
 * 
 * Returns `true` if the supplied `ContentRange` instance represents a satisfiable range.
 * 
 * @param range A `ContentRange` instance.
 * @returns `true` if `range` is a `ContentRange<InclRange>` instance, `false` otherwise.
 */
export function isSatisfiableRange(range: ContentRange): range is ContentRange<InclRange> {
    return range.range instanceof InclRange;
}

/**
 * Type guard function that narrows the supplied `ContentRange` instance to a `ContentRange<InclRange, number>` instance.
 * 
 * Returns `true` if the supplied `ContentRange` instance represents a complete-length range.
 * 
 * @param range A `ContentRange` instance.
 * @returns `true` if `range` is a `ContentRange<InclRange, number>` instance, `false` otherwise.
 */
export function isCompleteLengthRange(range: ContentRange): range is ContentRange<InclRange, number> {
    return isSatisfiableRange(range) && typeof range.length === 'number';
}


/**
 * Parses a HTTP `multipart/byteranges` response `body` delimited by the specified `boundary`.
 * Returns an array of `BodyPart<InclRange>` instances.
 * 
 * Throws if it cannot find part boundaries or valid (satisfiable) ranges.
 *   
 * @param body A `Buffer` instance.
 * @param boundary The body parts boundary.
 * @returns Array of body parts.
 */
export function parse(body: Buffer, boundary: string): BodyPart<InclRange>[] {
    const parts: BodyPart<InclRange>[] = [];

    boundary = '--' + boundary;

    let index = body.indexOf(boundary);

    if (index === -1) {
        throw new Error('Boundary not found');
    }

    while (index !== -1) {
        index += boundary.length;

        if (body[index] === 0x2D && body[index + 1] === 0x2D) {
            // --boudary-- indicates that no further body parts will follow
            break;
        }

        // discard transport padding and CRLF
        index = body.indexOf(crlf, index) + 2;

        if (index === -1) {
            throw new Error('Range begin not found');
        }

        let end = body.indexOf(crlfcrlf, index);

        if (end === -1) {
            throw new Error('Headers not found');
        }

        const headers = body.toString('ascii', index, end);
        const range = getPartContentRange(headers);

        if (!isSatisfiableRange(range)) {
            throw new Error('Unsatisfied range');
        }

        const type = getPartContentType(headers);
        const begin = end + 4;

        if (range.unit.toLowerCase() === 'bytes') {
            end = begin + range.range.end - range.range.start + 1;
            index = body.indexOf(boundary, end);
        } else {
            index = body.indexOf(boundary, begin);
            end = index - 2;
        }

        if (index === -1) {
            throw new Error('Range end not found');
        }

        const octets = body.subarray(begin, end);

        parts.push(new BodyPart(range, octets, type));
    }

    return parts;
}

/**
 * Type guard function that narrows the supplied `BodyPart[]` instance to a `BodyPart<InclRange, number>[]` instance.
 * 
 * Returns `true` if the supplied `BodyPart[]` instance represents an array of complete-length range parts.
 * 
 * @param range A `BodyPart[]` instance.
 * @returns `true` if `parts` is a `BodyPart<InclRange, number>[]` instance, `false` otherwise.
 */
export function isCompleteLengthRangeParts(parts: BodyPart[]): parts is BodyPart<InclRange, number>[] {
    return !parts.some(part => !isCompleteLengthRange(part.range));
}


examples/superagent.js
const { parse, getContentRange, BodyPart } = require('../build/byteranges');
const superagent = require('superagent');

superagent
    .get('http://www.columbia.edu/~fdc/picture-of-something.jpg')
    .set('Range', 'bytes=1-10,101-150')
    .buffer(true)
    .parse(superagent.parse['application/octet-stream'])
    .end((err, res) => {
        if (err) {
            console.log(err);
            return;
        }

        if (res.status !== 206) {
            return;
        }

        const contentType = res.headers['content-type'];
        if (!/multipart\/byteranges/i.test(contentType)) {
            const range = getContentRange(res.headers['content-range']);
            const part = new BodyPart(range, res.body, contentType);
            console.log(JSON.stringify(part));
            return;
        }

        const match = /boundary=(\w+)/i.exec(contentType);
        if (match !== null) {
            const boundary = match[1];
            const parts = parse(res.body, boundary);
            console.log(JSON.stringify(parts));
        }
    });
